acceptance/helpers.rb

```diff
@@ -66,114 +203,125 @@ def prepare_upgrade_with(initial_package_version_or_collection)
+
+              teardowns__clean_agents << lambda do
+                step "Teardown: Uninstall the puppet-agent package on agent #{host_to_info_s(agent)}" do
+                  if agent['platform'] =~ /windows/
+                    scp_to(agent, "#{SUPPORTING_FILES}/uninstall.ps1", "uninstall.ps1")
+                    on(agent, 'rm -rf C:/ProgramData/PuppetLabs')
+                    on(agent, 'powershell.exe -File uninstall.ps1 < /dev/null')
+                  else
+                    manifest_lines = []
+                    # Remove pc_repo:
+                    # Note pc_repo is specific to this module's manifests. This is
+                    # knowledge we need to clean from the machine after each run.
+                    if agent['platform'] =~ /debian|ubuntu/
+                      on(agent, puppet('module', 'install', 'puppetlabs-apt'), acceptable_exit_codes: [0])
+                      manifest_lines << 'include apt'
+                      manifest_lines << "apt::source { 'pc_repo': ensure => absent, notify => Package['puppet-agent'] }"
+                    elsif agent['platform'] =~ /fedora|el|centos/
+                      manifest_lines << "yumrepo { 'pc_repo': ensure => absent, notify => Package['puppet-agent'] }"
+                    end
+
+                    manifest_lines << "file { ['/etc/puppet', '/etc/puppetlabs', '/etc/mcollective']: ensure => absent, force => true, backup => false }"
+
+                    if agent['platform'] =~ /^(osx|solaris)/
+                      # The macOS pkgdmg and Solaris sun providers don't support 'purged':
+                      manifest_lines << "package { ['puppet-agent']: ensure => absent }"
+                    else
+                      manifest_lines << "package { ['puppet-agent']: ensure => purged }"
+                    end
+
+                    on(agent, puppet('apply', '-e', %("#{manifest_lines.join("\n")}"), '--no-report'), acceptable_exit_codes: [0, 2])
+                  end
+                end
+              end
             end
           end
 
-          step 'Clear SSL and stop firewalls' do
-            block_on(agents_only) do |agent|
+          step '(Agents) Clear SSL and stop firewalls' do
+            agents_only.each do |agent|
               ssldir = puppet_config(agent, 'ssldir').strip
               on(agent, "rm -rf '#{ssldir}'/*") # Preserve the directory itself, to keep permissions
 
               stop_firewall_with_puppet_on(agent)
             end
           end
 
-          step 'Sign certs' do
-            master_fqdn = on(master, facter('fqdn')).stdout.strip
-            master_hostname = on(master, facter('hostname')).stdout.strip
-            master_conf = { 'main' => {
-                'autosign'      => true,
-                'dns_alt_names' => "puppet,#{master_hostname},#{master_fqdn}",
-                'verbose'       => true,
-                'daemonize'     => true,
-            }}
-
-            with_puppet_running_on(master, master_conf) do
```

Comment:
Another thing I noticed was that `with_puppet_running_on` spent a lot of time restarting puppetserver and it also cluttered the output when `BEAKER_debug` was set. I based the cert-signing on what's in [ValidateSignCert](https://github.com/puppetlabs/beaker-puppet/blob/master/setup/common/040_ValidateSignCert.rb) instead since that avoids the restart.

Reply:
Do you have to restart puppetserver after clearing out old certs? I don't actually know if that's the case but if it is we need to ensure that's happening.

Reply:
Asked in the server room just now and it looks like the answer is "no".